include/bluetooth/rpc.rst

```diff
@@ -0,0 +1,70 @@
+.. _ble_rpc:
+
+Bluetooth Low Energy Remote Procedure Call
+##########################################
+
+.. contents::
+   :local:
+   :depth: 2
+
+The nRF Connect SDK supports Bluetooth Low Energy (LE) stack serialization.
+The full Bluetooth LE stack can run on another device or CPU, such as the nRF5340 DK network core using :ref:`nrfxlib:nrf_rpc`.
+
+.. note::
+   The nRF Connect SDK currently supports serialization of the :ref:`zephyr:bt_gap` and the :ref:`zephyr:bluetooth_connection_mgmt`.
+
+Network core
+************
+
+The :ref:`ble_rpc_host` sample is designed specifically to enable the Bluetooth LE stack functionality on a remote MCU (for example, the nRF5340 network core) using the :ref:`nrfxlib:nrf_rpc`.
+This sample can be programmed to the network core to run standard Bluetooth Low Energy samples on nRF5340.
+You can choose whether to use the SoftDevice Controller or the Zephyr Bluetooth LE Controller for this sample.
+
+Application core
+****************
+
+Bluetooth sample which are running on Application sample must be built with additional configuration.
+When building samples for the application core, enable the :option:`CONFIG_BT_RPC` to run Bluetooth LE stack on the network core.
+This option also causes that the :ref:`ble_rpc_host` is built automatically as a child image.
+For more details see: :ref:`ug_nrf5340_building`.
+
+Open a command prompt in the build folder of the application sample and enter the following command to build the application for the application core, with :ref:`ble_rpc_host` as child image:
+
+.. code-block:: console
+
+   west build -b nrf5340dk_nrf5340_cpuapp -- -DCONFIG_BT_RPC=y
```

Comment:
This is expected behavior since we supports at this moment only GAP and connection API. The GATT API is in a progress and will be introduced in separated PRs. You can build the `beacon` sample out of the box and `peripheral_sc_only` when you adjust the `rpc_host` sample configuration. This limitations are described in the documentation.

Reply:
> This is expected behavior since we supports at this moment only GAP and connection API. The GATT API is in a progress and will be introduced in separated PRs.

What is current status of this PR ?
I still see failures, just different now:
```
west build -b nrf5340dk_nrf5340_cpuapp --pristine=always samples/bluetooth/peripheral_cts_client/ -- -DCONFIG_BT_RPC=y
.....
[153/225] Building C object modules/nrf/subsys/nrf_rpc/CMakeFiles/..__nrf__subsys__nrf_rpc.dir/rp_ll.c.obj
FAILED: modules/nrf/subsys/nrf_rpc/CMakeFiles/..__nrf__subsys__nrf_rpc.dir/rp_ll.c.obj 
ccache /opt/gcc-arm-none-eabi-7-2018-q2-update/bin/arm-none-eabi-gcc -DBUILD_VERSION=zephyr-v2.5.0-4243-gdbf1094f3d56 -DKERNEL -DNRF5340_XXAA_NETWORK -DUSE_PARTITION_MANAGER=1 -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR__=1 -I../../drivers/mpsl/clock_control -I/projects/github/ncs/zephyr/include -Izephyr/include/generated -I/projects/github/ncs/zephyr/soc/arm/nordic_nrf/nrf53 -I/projects/github/ncs/zephyr/subsys/bluetooth -I../../include -I../../lib/multithreading_lock/. -I../../subsys/bluetooth/controller/. -I../../subsys/nrf_rpc/include -I/projects/github/ncs/modules/hal/cmsis/CMSIS/Core/Include -I/projects/github/ncs/modules/hal/nordic/nrfx -I/projects/github/ncs/modules/hal/nordic/nrfx/drivers/include -I/projects/github/ncs/modules/hal/nordic/nrfx/mdk -I/projects/github/ncs/zephyr/modules/hal_nordic/nrfx/. -Imodules/libmetal/libmetal/lib/include -I/projects/github/ncs/modules/lib/open-amp/open-amp/lib/include -I/projects/github/ncs/modules/lib/tinycbor/zephyr/../include -I/projects/github/ncs/modules/crypto/tinycrypt/lib/include -I/projects/github/ncs/nrfxlib/mpsl/include -I/projects/github/ncs/nrfxlib/softdevice_controller/include -I/projects/github/ncs/nrfxlib/nrf_rpc/include -I/projects/github/ncs/find-my/include -isystem /projects/github/ncs/zephyr/lib/libc/minimal/include -isystem /opt/gcc-arm-none-eabi-7-2018-q2-update/bin/../lib/gcc/arm-none-eabi/7.3.1/include -isystem /opt/gcc-arm-none-eabi-7-2018-q2-update/bin/../lib/gcc/arm-none-eabi/7.3.1/include-fixed -Os -imacros /projects/github/ncs/nrf/build/rpc_host/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -fdiagnostics-color=always -mcpu=cortex-m33+nodsp -mthumb -mabi=aapcs -imacros /projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT modules/nrf/subsys/nrf_rpc/CMakeFiles/..__nrf__subsys__nrf_rpc.dir/rp_ll.c.obj -MF modules/nrf/subsys/nrf_rpc/CMakeFiles/..__nrf__subsys__nrf_rpc.dir/rp_ll.c.obj.d -o modules/nrf/subsys/nrf_rpc/CMakeFiles/..__nrf__subsys__nrf_rpc.dir/rp_ll.c.obj -c /projects/github/ncs/nrf/subsys/nrf_rpc/rp_ll.c
/projects/github/ncs/nrf/subsys/nrf_rpc/rp_ll.c:16:10: fatal error: nrf_rpc_errno.h: No such file or directory
 #include <nrf_rpc_errno.h>
          ^~~~~~~~~~~~~~~~~
compilation terminated.
```

Reply:
I will fix that. I put wrong commit when I was solving the west.yaml conflicts.

Reply:
Fixed, but as @KAGA164 explained above, you will not be able to build this sample since it is using GATT API that is not serialized yet. You can try `beacon`.

We are currently focused on implementing GATT API serialization. It will be our second step of bluetooth API serialization planned for the next release. Then, you will be able to compile this sample.

Reply:
>  It will be our second step of bluetooth API serialization planned for the next release.

That was not how I understood this comment:
> This is expected behavior since we supports at this moment only GAP and connection API. The GATT API is in a progress and will be introduced in separated PRs. 

Nothing mentioned about this not working until next release.

Reading the docs doesn't make it clear to me that only beacon sample is supported.
> The nRF Connect SDK supports Bluetooth Low Energy (LE) stack serialization.
> The full Bluetooth LE stack can run on another device or CPU, such as the nRF5340 DK network core using :ref:`nrfxlib:nrf_rpc`.

so how are users expected to know what apps can build and run when having `-DCONFIG_BT_RPC=y` ?

Please ensure that samples that are not yet supporting `BT_RPC` cannot enable this feature.

Reply:
What is currently supported is mentioned in either general rpc.rst doc here

https://github.com/nrfconnect/sdk-nrf/pull/4161/files#diff-3bbad1e03fb677c87e7416ca25bb6bfedca633d179eebdf582dadc0d954b7ba4R13 

or in sample doc:

https://github.com/nrfconnect/sdk-nrf/pull/4161/files#diff-fd9f07b4851149f608640ad6a8aa95509afe3e71aa0e0e5d2d309c54a0f53db6R12

Reply:
> What is currently supported is mentioned in either general rpc.rst doc here
> 
> https://github.com/nrfconnect/sdk-nrf/pull/4161/files#diff-3bbad1e03fb677c87e7416ca25bb6bfedca633d179eebdf582dadc0d954b7ba4R13

Two problems with that:
1. This is just a note, not a warning, like `Only supported for ......, samples using other Bluetooth LE features, such as GATT are currently not supported`
2. The two links that users are directed to are those:
    https://docs.zephyrproject.org/latest/reference/bluetooth/gap.html
    https://docs.zephyrproject.org/latest/reference/bluetooth/connection_mgmt.html
    How will this help the user to determine if a particular sample is supported or not ?

Please safeguard the `BT_RPC` so it cannot be enabled when unsupported features are enable.

For example like this:
```
config BT_RPC
    bool "Bluetooth over RPC"
    depends on !BT_GATT_CLIENT
    ....
```

Reply:
as example, reworking a little on the Kconfig makes it possible to get this warning if building a sample the uses GATT client:
```
warning: BT_RPC (defined at /projects/github/ncs/nrf/subsys/bluetooth/rpc/Kconfig:7) was assigned
the value 'y' but got the value 'n'. Check these unsatisfied dependencies: (!BT_GATT_CLIENT) (=n).
See http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_BT_RPC.html and/or look up BT_RPC
in the menuconfig/guiconfig interface. The Application Development Primer, Setting Configuration
Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful too.
```

Reply:
>This is just a note, not a warning, like Only supported for ......, samples using other Bluetooth LE features, such as GATT are currently not supported

Ok, I can replace note with a warning similar to above.

> The two links that users are directed to are those:
> https://docs.zephyrproject.org/latest/reference/bluetooth/gap.html
> https://docs.zephyrproject.org/latest/reference/bluetooth/connection_mgmt.html
> How will this help the user to determine if a particular sample is supported or not ?

Simply, if sample uses BLE features other than listed, then it is not supported.

Safeguard as you described above is just a partial solution. It will work for GATT client, but there are other features that cannot be safeguarded this way, e.g. GATT server.

Reply:
I wanted to add `depends on !BT_GATT_CLIENT` as you suggested, but now all the samples are failing because of dependency loop. At first look, I don't see a good point where to break this loop, so I think that documentation must be enough for the user right now.

How did you get the warning that you shown. Did you made more changes than you provided in your example?

Reply:
Based on @peknis suggestion, I had to change `warning` in the documentation back to `note` because of Nordic's technical documentation style and usage guide.

Reply:
> I wanted to add depends on !BT_GATT_CLIENT as you suggested, but now all the samples are failing because of dependency loop

Probably because you should turn some of your Kconfigs around.
I notice that BT_RPC_HOST and BT_RPC_CLIENT tries to `select BT`, which in my opinion is wrong.
Having any kind of sample, and then enable BT_RPC and suddenly get Bluetooth enable sound wrong to me.
But having a Bluetooth sample, where BT is already enabled can of course enable BT_RPC also.

So it appears that there should be depends in places where you are currently using select.

Have discussed a short term solution with @grochu.

To scope the current solution to only those sample currently supported:
Place `<sample>/Kconfig` with
```
config SUPPORT_BT_RPC
    bool
    default y

source "Kconfig.zephyr"
```

and in `subsys/bluetooth/rpc/Kconfig`
```
config SUPPORT_BT_RPC
    bool

config BT_RPC
    .....
    depends on SUPPORT_BT_RPC
```

Reply:
I have one concern about that. This solves issue with samples, but remember that our users do their own applications, so they have to also create a Kconfig file in their projects and copy-paste code from your first code block.
This have to be documented, so users can use it in their applications.

Maybe making `SUPPORT_BT_RPC` visible to the user will be more convenient to him?

Reply:
Done.

Reply:
> This solves issue with samples, but remember that our users do their own applications, so they have to also create a Kconfig file in their projects and copy-paste code from your first code block.

correct, but in this case the Kconfig file is located with the sample, so in this case, if they copy beacon folder then they get this file as well.

If they copy peripheral_hrs, they don't get it, and they don't accidentally set this setting.
Of course, the best is to untangle the Kconfigs to avoid the circular deps, and ensure the needed symbols is in place, and that can be done as the GATT support is implemented.

Reply:
I put changes in the last commit.

@tejlmand, Is the PR acceptable for you now?

Reply:
> I wanted to add depends on !BT_GATT_CLIENT as you suggested, but now all the samples are failing because of dependency loop. 

I took an extra look at you Kconfig, and this should fix the issue, and actually remove the need for:
```
config SUPPORT_BT_RPC
       bool
       default y

source "Kconfig.zephyr"
```
https://github.com/KAGA164/fw-nrfconnect-nrf/pull/5

With the referenced PR then it is clear that BT_RPC depends on BT_GATT or BT_NRF_SERVICES not being enable, and thus ensuring a user cannot enable BT_RPC, for example when building HIDS Keyboard.

A user trying to enable BT_RPC anyway, will see the following warning:
```
$ west build -b nrf5340dk_nrf5340_cpuapp --pristine=always samples/bluetooth/peripheral_hids_keyboard/ -- -DCONFIG_BT_RPC=y
.....

warning: BT_RPC (defined at /projects/github/ncs/nrf/subsys/bluetooth/rpc/Kconfig:7) was assigned
the value 'y' but got the value 'n'. Check these unsatisfied dependencies: (!BT_NRF_SERVICES) (=n).
See http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_BT_RPC.html and/or look up BT_RPC
in the menuconfig/guiconfig interface. The Application Development Primer, Setting Configuration
Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful too.

```

Next time I point out an issue, please get in touch immediately if you don't know how to solve it, instead of not doing anything.

Reply:
I think the problem here was a little different than not knowing how to fix the problem. I just didn't agree with you that we have any problem here at the beginning so I responded to your comments with something like "This is expected behaviour". If you said that from your point of view it should be handled from Kconfig level or that documentations here might be confusing for the user then I would probably agree with your comments and we would end with similar solution like you propose immediately :)

Reply:
Hi @tejlmand, I think we discussed the non-technical part of handling this PR offline, se let's not escalate this kind of arguments here e.g. 

> Next time I point out an issue, please get in touch immediately if you don't know how to solve it, instead of not doing anything.

Especially, that the PR had been hanging without review for enough time earlier, and I don't think anyone was delaying responding to the comments from the authors' side.